### ISSUE TITLE
Synced Realm collection notifications not being received

### ISSUE THREAD

**Issue by duncangroenewald** [realm/realm-swift#7548]
### How frequently does the bug occur?

Sometimes

### Description

We expect to be able to observe collection changes using the following code:

class ViewController: NSViewController {

....
  var changeNoficationToken: NotificationToken?

func register(){

`        self.changeNotificationToken = changes?.observe { [weak self] (changes: RealmCollectionChange) in
            
            DebugLog("UserChange changed")
            
            switch changes {
            case .initial:
                
                self?.tableView.reloadData()
                break
                
            case .update(_,  _,  _,  _):
                
                self?.tableView.reloadData()
                break
                
            case .error(let error):
                // An error occurred while opening the Realm file on the background worker thread
                fatalError("\(error)")
            }
        }`
}

}

However for some reason this is not getting called when changes are made by other remote synced clients.  But doing a manual tableView.reloadData() does show the changed data.

Are there some circumstances where these notifications will not be sent - and am I correct in assuming that if register() is called on the main thread then the callback will be called on the main thread.

### Stacktrace & log output

```shell
N/A
```

### Can you reproduce the bug?

Yes, sometimes

### Reproduction Steps

Currently a production system

### Version

RealmSwift v10.19.0

### What SDK flavour are you using?

MongoDB Realm (i.e. Sync, auth, functions)

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

macOS 12.0.1

### Build environment

Xcode version: 13.1
Dependency manager and version: none - using downloaded RealmSwift prebuilt packages for Xcode 13.1

**Comment by duncangroenewald** [comment]
So this seems a bit weird because intermittently when opening the synced realm the call to `realm.syncSession?.addProgressNotification(...)` will return a non-nil value but the callback will subsequently never be called.  And when this occurs the collection notifications are delivered correctly.

However whenever the callback is called the collection notifications don't seem to be called.

It doesn't seem to help to not add a progress notification block.  It seems that opening the realm file will intermittently deliver the collection change notifications, but more frequently than not it won't.

**Comment by duncangroenewald** [comment]
OK I think I found the problem.

When we open the synced realm the first time when the app gets launched by the user we do so from a background thread to avoid any issues with any realm-core database file format upgrades that might need to be run.

If this upgrade needs to be run then this causes the MAIN THREAD to block until the database file format upgrade has completed.  This is obviously undesirable because the UI hangs.

If I remove this first time call to Realm() and do it from the MAIN THREAD then this issue of receiving change notifications appears to now go away.

Strange that the Intel macOS users don't appear to have reported any problems with the UI not updating when other clients update the data.  

So I guess the question is should it matter that the first time Realm() is called it is called from a background thread ?  Is this a bug perhaps ?

**Comment by jsflax** [comment]
@duncangroenewald can you send us a project that demonstrates this behaviour? Thanks.

**Comment by duncangroenewald** [comment]
@jsflax - I can try and create one for you.  However I just discovered that the cause of the problem might have something to do with the fact that our initial call is like this below (pseudocode).

Not 100% sure here but is seems that because the realm returned by the initial call to Realm() is not being retained this may be causing a problem for the realm-core setup code.

I am speculating a bit here but I noticed the same issue with a test app where the sync callbacks would never be called if the realm returned from the initial call to Realm() was not retained.  I need to investigate further to be sure though but thought that might be worth mentioning.  

I will try and reproduce with a test app for you.

func initRealm(completion: ()->Void) {

DispatchQueue.global().async {

   // Open the synced realm the first time on a background thread so
   // any blocking because of file format migrations done by realm-core
   // don't block the UI 
   let realm = Realm(config)

   DispatchQueue.main.async {
   
      // Now open the main thread Realm used by the UI in the app
      Global.shared.realmMain = Realm(config)
      
      completion()
   }

}
}
